lib/ide/ast/impl/src/known.rs

```diff
@@ -0,0 +1,130 @@
+//! This module provides KnownAst<T> wrapper over Ast that allows expressing that we already
+//! know what `Shape` variant is being stored within this `Ast` node.
+
+use crate::prelude::*;
+
+use crate::Ast;
+use crate::Shape;
+
+
+
+// =================
+// === Known AST ===
+// =================
+
+/// Wrapper for an AST node of known shape type that we can access.
+/// Use `TryFrom<&Ast>` to obtain values.
+///
+/// Provides `Deref` implementation that allows accessing underlying shape `T` value.
+#[derive(Derivative)]
+#[derivative(Clone(bound=""))]
+#[derive(Debug)]
+pub struct KnownAst<T>(Ast, PhantomData<T>);
+
+impl<T> KnownAst<T> {
+    /// Checks if the shape of given Ast node is compatible with `T`.
+    /// If yes, returns Ok with Ast node wrapped as KnownAst.
+    /// Otherwise, returns an error.
+    pub fn try_new<E>(ast:Ast) -> Result<KnownAst<T>,E>
+    where for<'t> &'t Shape<Ast>: TryInto<&'t T, Error=E> {
+        if let Some(error_matching) = ast.shape().try_into().err() {
+            Err(error_matching)
+        } else {
+            Ok(KnownAst(ast,default()))
+        }
+    }
+}
+
+impl<T,E> Deref for KnownAst<T>
+where for<'t> &'t Shape<Ast> : TryInto<&'t T,Error=E>,
+                           E : Debug, {
+    type Target = T;
+    fn deref(&self) -> &Self::Target {
+        let result = self.0.shape().try_into();
```

Comment:
I really dont like using `.0`. It always feel hacky. Luckily, you can implement `HasContent` and `RefContent` from prelude to have method `.content()` here :) (and many more!)

Reply:
If one does not like `.0` you can write:
```rust
let KnownAst(ast,_) = self
```
It's more expressive. For me `self.content()` would look just, hmm, "ugly". If I would not know the code I would wonder "what is the content of Ast?". And when I alt+click to content I get to definition of content in some HasContent abstraction which explains exactly nothing.

That's why I'm against the abstractions like `HasContent`